### PR TITLE
Screen markers widgets

### PR DESCRIPTION
**Pull request by Jasper-ketelaar**
This feature allows you to attach screen markers to widgets by pressing shift when creating the screen marker. Also works on items.

screen markers

### CODE REVIEW

**Comment by psikoi** [comment]
Maybe show a translucent hover effect on the widget? this would help the player see what they're selecting

**Comment by DannysPVM** [comment]
You may have already seen the little checkstyle errors that Travis is reporting, but just in case here they are :)

```
[INFO] There are 3 errors reported by Checkstyle 8.3 with checkstyle.xml ruleset.
[ERROR] src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerKeyListener.java:[27,8] (imports) UnusedImports: Unused import - java.awt.event.KeyAdapter.
[ERROR] src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerKeyListener.java:[30,8] (imports) UnusedImports: Unused import - net.runelite.api.Client.
[ERROR] src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerKeyListener.java:[31,8] (imports) UnusedImports: Unused import - net.runelite.client.callback.ClientThread.
```

looks like it's just three unused imports, so in ScreenMarkerKeyListener.java remove these lines:

```java
(Line 27) import java.awt.KeyAdapter;
(Line 30 ) import net.runelite.api.Client;
(Line 31) import net.runelite.client.callback.ClientThread;
```

**Comment by Jasper-ketelaar** [comment]
Alright fixed those imports and added an indicator to see what you are selecting.

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerCreationOverlay.java [hunk 1]
```diff
@@ -39,7 +40,7 @@
 	private ScreenMarkerPlugin plugin;
 
 	@Inject
-	private ScreenMarkerCreationOverlay(final ScreenMarkerPlugin plugin)
+	private ScreenMarkerCreationOverlay(final ScreenMarkerPlugin plugin, final Client client)
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerCreationOverlay.java]: client is not used

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerCreationOverlay.java [hunk 2]
```diff
@@ -71,6 +72,6 @@ public Dimension render(Graphics2D graphics)
 
 	private Stroke createStripedStroke(int thickness)
 	{
-		return new BasicStroke(thickness, BasicStroke.CAP_BUTT, BasicStroke.JOIN_BEVEL, 0, new float[]{9}, 0);
+		return new BasicStroke(thickness, BasicStroke.CAP_BUTT, BasicStroke.JOIN_BEVEL, 0, new float[] {9}, 0);
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerCreationOverlay.java]: The whitespace change here is unnecessary imo

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerOverlay.java [hunk 3]
```diff
@@ -38,11 +38,13 @@
 {
 	@Getter
 	private final ScreenMarker marker;
+	private final ScreenMarkerPlugin plugin;
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerOverlay.java]: Not used anywhere

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java [hunk 4]
```diff
@@ -76,15 +82,25 @@
 	@Inject
 	private MouseManager mouseManager;
 
+	@Inject
+	private KeyManager keyManager;
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java]: Not used anywhere

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java [hunk 5]
```diff
@@ -174,18 +232,141 @@ public void startCreation(Point location)
...
 		);
 
+		creatingScreenMarker = true;
+
 		// Set overlay creator bounds to current position and default size
 		startLocation = location;
 		overlay.setPreferredLocation(location);
 		overlay.setPreferredSize(DEFAULT_SIZE);
-		creatingScreenMarker = true;
+
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java]: Delete empty new line

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
+package net.runelite.client.plugins.screenmarkers;
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java]: Add license header

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+import net.runelite.api.widgets.Widget;
+import net.runelite.api.widgets.WidgetItem;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayLayer;
+import net.runelite.client.ui.overlay.OverlayPosition;
+import net.runelite.client.ui.overlay.OverlayPriority;
+
+public class ScreenMarkerWidgetHighlight extends Overlay
+{
+
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java]: Delete empty new line

File: runelite-mixins/src/main/java/net/runelite/mixins/RSWidgetMixin.java [hunk 8]
```diff
@@ -38,16 +38,16 @@
...
-import static net.runelite.client.callback.Hooks.log;
 import net.runelite.rs.api.RSClient;
 import net.runelite.rs.api.RSHashTable;
 import net.runelite.rs.api.RSNode;
 import net.runelite.rs.api.RSWidget;
+import static net.runelite.api.widgets.WidgetInfo.TO_CHILD;
+import static net.runelite.api.widgets.WidgetInfo.TO_GROUP;
+import static net.runelite.client.callback.Hooks.deferredEventBus;
+import static net.runelite.client.callback.Hooks.eventBus;
+import static net.runelite.client.callback.Hooks.log;
```
Comment (jplsek) [inline runelite-mixins/src/main/java/net/runelite/mixins/RSWidgetMixin.java]: Fix imports

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java (hunk 5 above)
Comment (Kamielvf) [reply runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java]: Why was this moved in the first place? Seems unnecessary.

File: runelite-client/src/main/java/net/runelite/client/plugins/idlenotifier/IdleNotifierPlugin.java [hunk 9]
```diff
@@ -25,13 +25,12 @@
  */
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/idlenotifier/IdleNotifierPlugin.java]: Revert this file

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2018, Kamiel, <https://github.com/Kamielvf>
+ * Copyright (c) 2018, Adam <Adam@sigterm.info>
+ * Copyright (c) 2018, Psikoi <https://github.com/psikoi>
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java]: The copyright should be attributed to just you

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+import javax.inject.Inject;
+import net.runelite.api.Client;
+import net.runelite.api.widgets.Widget;
+import net.runelite.api.widgets.WidgetItem;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayLayer;
+import net.runelite.client.ui.overlay.OverlayPosition;
+import net.runelite.client.ui.overlay.OverlayPriority;
+
+public class ScreenMarkerWidgetHighlight extends Overlay
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java]: Remove public

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+{
+	private static final int MAX_SIZE = 8000;
+
+	private ScreenMarkerPlugin plugin;
+	private Client client;
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java]: Make final

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayLayer;
+import net.runelite.client.ui.overlay.OverlayPosition;
+import net.runelite.client.ui.overlay.OverlayPriority;
+
+public class ScreenMarkerWidgetHighlight extends Overlay
+{
+	private static final int MAX_SIZE = 8000;
+
+	private ScreenMarkerPlugin plugin;
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerWidgetHighlight.java]: Make final

File: runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java [hunk 14]
```diff
@@ -163,6 +178,45 @@ public void setMouseListenerEnabled(boolean enabled)
...
+			pluginPanel.getSelectedFillColor(),
+			true
+		);
+
+		startLocation = new Point(bounds.x, bounds.y);
+		overlay.setPreferredLocation(startLocation);
+		overlay.setPreferredSize(bounds.getSize());
+	}
+
+
```
Comment (jplsek) [inline runelite-client/src/main/java/net/runelite/client/plugins/screenmarkers/ScreenMarkerPlugin.java]: Remove new line

**Pull request closed without merging**